Mathlib/Data/Nat/Factorization/Root.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.Order.Floor.Div
+import Mathlib.Data.Nat.Factorization.Basic
+
+/-!
+# Roots of natural numbers, rounded up and down
+
+This file defines the flooring and ceiling root of a natural number. The `n`-th flooring/ceiling
+root of `a` is the natural number whose `p`-adic valuation is the floor/ceil of the `p`-adic
+valuation of `a`.
+
+For example the `2`-nd flooring and ceiling roots of `2^3 * 3^2 * 5` are `2 * 3` and `2^2 * 3 * 5`
+respectively. Note this is **not** the `n`-th root of `a` as a real number, rounded up or down.
+
+Those operations are respectively the right and left adjoints to the map `a ↦ a ^ n`.
```

Comment:
I guess you have to specify the ordering. You mean the divisibility ordering, right? Because for the natural ordering, I suppose that `\lfloor sqrt{n} a \rfloor` (resp `ceil`) would be the adjoints?